third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -528,8 +537,73 @@ impl InterpreterImpl {
                         loader,
                         function,
                         frame_cache,
+                        ClosureMask::empty(),
+                        vec![],
                     )?;
                 },
+                ExitCode::CallClosure(_sig_idx) => {
+                    // TODO(#15664): runtime check whether closure value on stack is same as sig_idx
+                    let (fun, captured) = self
+                        .operand_stack
+                        .pop_as::<Closure>()
+                        .map_err(|e| set_err_info!(current_frame, e))?
+                        .unpack();
+                    let lazy_function = LazyLoadedFunction::expect_this_impl(fun.as_ref())
+                        .map_err(|e| set_err_info!(current_frame, e))?;
+                    let mask = lazy_function.closure_mask();
+                    let function = lazy_function
+                        .with_resolved_function(module_storage, |f| Ok(f.clone()))
+                        .map_err(|e| set_err_info!(current_frame, e))?;
+
+                    // In difference to regular calls, we skip visibility check.
+                    // It is possible to call a private function of another module via
+                    // a closure.
+
+                    // Charge gas
+                    let module_id = function.module_id().ok_or_else(|| {
+                        let err =
+                            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                                .with_message(
+                                    "Failed to get native function module id".to_string(),
+                                );
+                        set_err_info!(current_frame, err)
+                    })?;
+                    gas_meter
+                        .charge_call(
+                            module_id,
+                            function.name(),
+                            self.operand_stack
+                                .last_n(function.param_tys().len())
+                                .map_err(|e| set_err_info!(current_frame, e))?,
+                            (function.local_tys().len() as u64).into(),
+                        )
+                        .map_err(|e| set_err_info!(current_frame, e))?;
+
+                    // Call function
+                    if function.is_native() {
```

Comment:
Ah capturing natives is indeed an interesting corner case here. Do you think we should even forbid such case from happening?